jrnl/plugins/pelican_markdown_exporter.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+# encoding: utf-8
+# Copyright (C) 2012-2021 jrnl contributors
+# License: https://www.gnu.org/licenses/gpl-3.0.html
+
+import os
+import re
+import sys
+
+from jrnl.color import ERROR_COLOR
+from jrnl.color import RESET_COLOR
+from jrnl.color import WARNING_COLOR
+
+from .text_exporter import TextExporter
+
+
+class PelicanMarkdownExporter(TextExporter):
+    """
+    This Exporter can convert entries and journals into Markdown formatted
+    text with YAML front matter, as may be used by Pelican (a static site
+    generator).
+    """
+
+    names = ["pelican-markdown"]
+    extension = "md"
+
+    @classmethod
+    def export_entry(cls, entry, to_multifile=True):
+        """Returns a markdown representation of a single entry, with YAML front matter."""
+        if to_multifile is False:
+            print(
+                f"{ERROR_COLOR}ERROR{RESET_COLOR}: Pelican Markdown export must \
+                be to individual files. Please specify a directory to export to.",
+                file=sys.stderr,
+            )
+            return
+
+        date_str = entry.date.strftime(entry.journal.config["timeformat"])
+        body_wrapper = "\n" if entry.body else ""
+        body = body_wrapper + entry.body
+
+        tagsymbols = entry.journal.config["tagsymbols"]
+        # see also Entry.Entry.rag_regex
+        multi_tag_regex = re.compile(fr"(?u)^\s*([{tagsymbols}][-+*#/\w]+\s*)+$")
+
+        """Increase heading levels in body text"""
+        newbody = ""
+        heading = "#"
+        previous_line = ""
+        warn_on_heading_level = False
+        for line in body.splitlines(True):
+            if re.match(r"^#+ ", line):
+                """ATX style headings"""
+                newbody = newbody + previous_line + heading + line
+                if re.match(r"^#######+ ", heading + line):
+                    warn_on_heading_level = True
+                line = ""
+            elif re.match(r"^=+$", line.rstrip()) and not re.match(
+                r"^$", previous_line.strip()
+            ):
+                """Setext style H1"""
+                newbody = newbody + heading + "# " + previous_line
+                line = ""
+            elif re.match(r"^-+$", line.rstrip()) and not re.match(
+                r"^$", previous_line.strip()
+            ):
+                """Setext style H2"""
+                newbody = newbody + heading + "## " + previous_line
+                line = ""
+            elif multi_tag_regex.match(line):
+                """Tag only lines"""
+                line = ""
+            else:
+                newbody = newbody + previous_line
+            previous_line = line
+        newbody = newbody + previous_line  # add very last line
```

Comment:
I like the commenting and general readability of this section, but it looks like a good candidate to be broken out into its own function. The overall flow of the exporter would be easier to read, and the specific  function could be unit-tested. What do you think?

Reply:
It is likely a good section to be broken out into a function, except that it isn't used elsewhere and so hasn't already.

In any case, this is the code as it has been since 2014, and significant refactoring of the plugin is beyond the intended scope of this PR.